### PR TITLE
Support plugin event

### PR DESCRIPTION
**Pull request by ben181231**
connect #199

### CODE REVIEW

**Comment by oursky-travis** [comment]
@ben181231, thanks for your PR! By analyzing the annotation information on this pull request, we identified @rickmak to be a potential reviewer

**Comment by ben181231** [comment]
@cheungpat Please help to review this PR.

File: pkg/server/plugin/event/event.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package event
+
+import "github.com/skygeario/skygear-server/pkg/server/logging"
+
+var log = logging.LoggerEntry("plugin-evnet")
```
Comment (cheungpat) [inline pkg/server/plugin/event/event.go]: s/plugin-evnet/plugin-event/

File: pkg/server/plugin/exec/process_test.go [hunk 2]
```diff
@@ -123,6 +117,12 @@ func TestRun(t *testing.T) {
 			So(string(out), ShouldEqual, `"op hello:world"`)
 		})
 
+		Convey("event", func() {
+			out, err := transport.SendEvent("foo:bar", []byte{})
```
Comment (cheungpat) [inline pkg/server/plugin/exec/process_test.go]: prefer foo-bar here

File: pkg/server/plugin/plugin.go [hunk 3]
```diff
@@ -116,55 +125,123 @@ type InitContext struct {
...
 	return true
 }
 
-// Init instantiates a plugin. This sets up hooks and handlers.
-func (p *Plugin) Init(context *InitContext) {
-	p.transport.SetInitHandler(func(out []byte, err error) error {
-		if err != nil {
-			panic(fmt.Sprintf("Unable to get registration info from plugin. Error: %v", err))
+// SendEvents sends event to all plugins
+func (c *Context) SendEvents(name string, data []byte, async bool) {
```
Comment (cheungpat) [inline pkg/server/plugin/plugin.go]: I would prefer `SendEvent`

Comment (cheungpat) [inline pkg/server/plugin/plugin.go]: need to mention explicitly in the comment that `async` being `false` means the event messages are sent to each plugin one after or another, or that event messages are sent to all plugins at once, then wait for all operations to complete

Comment (ben181231) [inline pkg/server/plugin/plugin.go]: Okay. I will update the comment.

**Comment by cheungpat** [comment]
Just adding a note that we still need schema-change event in RecordSaveHandler.

**Comment by ben181231** [comment]
@cheungpat The PR is updated.

File: pkg/server/handler/record.go [hunk 4]
```diff
@@ -345,6 +347,17 @@ func (h *RecordSaveHandler) Handle(payload *router.Payload, response *router.Res
 		results = append(results, result)
 	}
 
+	if resp.SchemaUpdated && h.EventSender != nil {
```
Comment (cheungpat) [inline pkg/server/handler/record.go]: move this section out of this function scope (incl or excl the if statement

File: pkg/server/handler/schema.go [hunk 5]
```diff
@@ -51,8 +51,8 @@ type schemaField struct {
 	TypeName string `mapstructure:"type" json:"type"`
 }
 
-func (resp *schemaResponse) Encode(data map[string]skydb.RecordSchema) {
-	resp.Schemas = make(map[string]schemaFieldList)
+func encodeRecordSchemas(data map[string]skydb.RecordSchema) map[string]schemaFieldList {
```
Comment (cheungpat) [inline pkg/server/handler/schema.go]: Move this to recordutil or (new) schemautil.

Basically we try not to have different handler files (record and schema) with code calling code in other file.

File: pkg/server/handler/schema.go [hunk 6]
```diff
@@ -66,8 +66,23 @@ func (resp *schemaResponse) Encode(data map[string]skydb.RecordSchema) {
...
 		}
 		sort.Sort(fieldList)
-		resp.Schemas[recordType] = fieldList
+		schemaMap[recordType] = fieldList
 	}
+
+	return schemaMap
+}
+
+func sendSchemaChangedEvent(
```
Comment (cheungpat) [inline pkg/server/handler/schema.go]: good idea to move this in a recordutil/schemautil file

File: pkg/server/skydb/pq/schema_test.go [hunk 7]
```diff
@@ -28,7 +28,7 @@ func TestExtend(t *testing.T) {
 		db := c.PublicDB().(*database)
 
 		Convey("return Resemble RecordSchema on second call", func() {
-			err := db.Extend("note", skydb.RecordSchema{
+			_, err := db.Extend("note", skydb.RecordSchema{
```
Comment (cheungpat) [inline pkg/server/skydb/pq/schema_test.go]: This test file tests `Extend`. At the very least you want to test the boolean return value from the `Extend` function.

File: pkg/server/handler/schema.go (hunk 5 above)
Comment (ben181231) [inline pkg/server/handler/schema.go]: I see.

**Comment by cheungpat** [comment]
good job benlei!

**Pull request merged**